enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,298 @@
+---
+title: Managing Control Plane machines
+authors:
+  - enxebre
+reviewers:
+  - hexfusion
+  - jeremyeder
+  - abhinavdahiya
+  - joelspeed
+  - smarterclayton
+  - derekwaynecarr
+approvers:
+  - TBD
+
+creation-date: 2020-04-02
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+  - https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md
+  - https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md
+  - https://github.com/openshift/enhancements/blob/master/enhancements/etcd/disaster-recovery-with-ceo.md
+  - https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/auto-cert-recovery.md
+  - https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md
+  - https://github.com/openshift/cluster-kube-scheduler-operator
+  - https://github.com/openshift/cluster-kube-controller-manager-operator
+  - https://github.com/openshift/cluster-openshift-controller-manager-operator
+  - https://github.com/openshift/cluster-kube-apiserver-operator
+  - https://github.com/openshift/cluster-openshift-apiserver-operator
+replaces:
+superseded-by:
+---
+
+# Managing Control Plane machines
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+
+## Glossary
+
+Control Plane: The collection of stateless and stateful processes which enable a Kubernetes cluster to meet minimum operational requirements. This includes: kube-apiserver, kube-controller-manager, kube-scheduler, kubelet and etcd.
+
+## Summary
+
+This proposal outlines a solution to provide self managed Control Plane Machines:
+  - Ensures that Control Plane Machines are recreated on a deletion request at any time.
+  - Ensures that Control Plane Machines are auto repaired when a node goes unready (Machine Health Check).
+
+Particularly:
+- Proposes to introduce a new CRD `ControlPlane` which automates the adoption of any existing Control Plane Machine:
+      - The ControlPlane controller creates and manages a MachineSet to back each Control Plane Machine that is found at any time.
```

Comment:
yes.
That way we provide same ux than we do atm for recreating machines coming up with fresh config and the underlying ms can be leveraged for vertical upgrades.

Reply:
This seems risky.  What if the CP controller deletes the existing machine instead of the new one?

Reply:
The CP controller would not delete anything, it enforces a replica number on machineSet.
The machineSet scales back to the replica number we enforce honouring the deletion strategy that we enforce.

Reply:
As this would happen quickly, if a second Machine has been created/reconciled, ie, before the cloud provider instance has time to properly start, it wouldn't have a NodeRef yet. Will the MachineSet controller pick the Machine that doesn't have the NodeRef over the one that does have a NodeRef when scaling down? Do we have guarantees for that?

Reply:
>the deletion strategy that we enforce.

That'd be covered by newest policy behaviour https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/machine/v1beta1/machineset_types.go#L66-L69

Reply:
Then that's another thing that needs to be specified.  We have to constantly enforce that value as well.

Reply:
sure, I meant that to implicit here https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR179
Added more details here https://github.com/openshift/enhancements/pull/292/commits/465a731d628d66074b6f3867aa0b5d3eec640781